src/main/scala/org/allenai/pipeline/Pipeline.scala

```diff
@@ -1,75 +1,206 @@
+        persist(step, LineCollectionIo.json[T], stepName, suffix)
+    }
+
+    /** Persist a single object */
+    object Singleton {
+      def asText[T: StringSerializable: ClassTag](
+        step: Producer[T],
+        stepName: String = null,
+        suffix: String = ".txt"
+      )(): PersistedProducer[T, FlatArtifact] =
+        persist(step, SingletonIo.text[T], stepName, suffix)
+
+      def asJson[T: JsonFormat: ClassTag](
+        step: Producer[T],
+        stepName: String = null,
+        suffix: String = ".json"
+      )(): PersistedProducer[T, FlatArtifact] =
+        persist(step, SingletonIo.json[T], stepName, suffix)
+    }
+
+    /** Persist an Iterator */
+    object Iterator {
+      def asText[T: StringSerializable: ClassTag](
+        step: Producer[Iterator[T]],
+        stepName: String = null,
+        suffix: String = ".txt"
+      ): PersistedProducer[Iterator[T], FlatArtifact] =
+        persist(step, LineIteratorIo.text[T], stepName, suffix)
+
+      def asJson[T: JsonFormat: ClassTag](
+        step: Producer[Iterator[T]],
+        stepName: String = null,
+        suffix: String = ".json"
+      )(): PersistedProducer[Iterator[T], FlatArtifact] =
+        persist(step, LineIteratorIo.json[T], stepName, suffix)
+    }
+
+  }
+
+  /** Create a persisted version of the given Producer
+    * The producer is registered as a target of the pipeline, and will be computed
+    * when the pipeline is run.
+    * See Persist.Collection, Persist.Singleton, etc. utility methods above
+    * @param original the non-persisted Producer
+    * @param io the serialization format
+    * @param suffix a file suffix
+    * @return the persisted Producer
+    */
+  def persist[T, A <: Artifact: ClassTag](
+    original: Producer[T],
+    io: Serializer[T, A] with Deserializer[T, A],
+    name: String = null,
+    suffix: String = ""
+  ): PersistedProducer[T, A] = {
+    val stepName = Option(name).getOrElse(original.stepInfo.className)
+    val path = s"data/$stepName.${hashId(original, io)}$suffix"
+    persistToArtifact(original, io, createOutputArtifact[A](path), name)
+  }
+
+  def persistToUrl[T, A <: Artifact: ClassTag](
+    original: Producer[T],
+    io: Serializer[T, A] with Deserializer[T, A],
+    url: URI,
+    name: String = null
+  ): PersistedProducer[T, A] = {
+    persistToArtifact(original, io, artifactFactory.createArtifact[A](url), name)
+  }
+
+  /** Persist this Producer and add it to list of targets that will be computed when the pipeline is run */
+  def persistToArtifact[T, A <: Artifact](
+    original: Producer[T],
+    io: Serializer[T, A] with Deserializer[T, A],
+    artifact: A,
+    name: String = null
+  ): PersistedProducer[T, A] = {
+    val persisted = new ProducerWithPersistence(original, io, artifact)
+    val stepName = Option(name).getOrElse(original.stepInfo.className)
+    addTarget(stepName, persisted)
+  }
+
+  def persistCustom[T, P <: Producer[T], A <: Artifact: ClassTag](
```

Comment:
This is a substitute for overriding `Producer.persist`, which had some problems.  First, a `Producer` might want to be persisted different ways in different pipelines.  Second, when making a custom persister, one usually has different (tighter) type constraints on the `Artifact` and `ArtifactIo` to be used.  The `FaultTolerantRddMapper` is good example of where this turned out to be quite awkward.